Generating a documentation for tests

ifxfrancois opened the pull request:
This PR add the new option `--document-tests`  to rustdoc.
When this option is set:
*  the test modules are not filtered out by rustdoc when generating the documentation,
* the functions marked as `#[test]` are collected into a new category called `Tests` instead of `Functions`

**Note:** tests are often not marked as public so the option `--document-private-items` may be required to see the tests in the documentation.

A short example of the generated docs can be found [here](https://ifxfrancois.github.io/rust_doc_test_demo/target/doc/rust_doc_test_demo/tests/index.html).
# Motivation

During the development of embedded software, Quality or certifications teams need a quick overview of the tests of a project. Typical relevant information are what, how and why a test tests.

* What would be a short description of the test.
* How would be a short list of the important tests steps.
* Why would be references to the relevant detailed architecture or requirements documents.

This information can easily be written as markdown in the documentation of the test functions and the documentation be generated with the additional option. The generated documentation can then be provided to relevant teams.


# Limitations

There is no integration with cargo and integrations test can not be added to the same documentation as the rest of the crate documentation. 


<!-- homu-ignore:start -->
<!--
If this PR is related to an unstable feature or an otherwise tracked effort,
please link to the relevant tracking issue here. If you don't know of a related
tracking issue or there are none, feel free to ignore this.

This PR will get automatically assigned to a reviewer. In case you would like
a specific user to review your work, you can assign it to them by using

    r​? <reviewer name>
-->
<!-- homu-ignore:end -->

jieyouxu commented on src/librustdoc/clean/types.rs:
```diff
@@ -659,7 +659,10 @@ impl Item {
                 }
             }
-            ItemKind::FunctionItem(_) | ItemKind::MethodItem(_, _) | ItemKind::TyMethodItem(_) => {
+            ItemKind::FunctionItem(_)
+            | ItemKind::MethodItem(_, _)
+            | ItemKind::TyMethodItem(_)
+            | ItemKind::TestItem(_) => {
```
Modelling a `#[test] fn` as a distinct `ItemKind` seems a bit strange: `#[test] fn`s are still function items, I wonder if you can distinguish non-`#[test]` fns versus `#[test]` fns by looking at the existence/absence of the `#[test]` attribute.

But yeah, rendering docs for tests do seem useful. Though a question that immediately comes to my mind is that do test-affiliated module docs get rendered too? As in e.g. docs on `#[cfg(test)] mod tests;`. I would think that they do, but I haven't looked closely if they do in this implementation.

ifxfrancois replied:
Yes, they are rendered in this implementation though they are most likely not `pub` therefore the additional flag `--document-private-items` may be needed.

GuillaumeGomez replied:
I agree with @jieyouxu, but also I like this difference as it makes it simpler to know what's being manipulated...

As for `cfg`ed out items, unfortunately you will need to pass `--cfg test` to have them (alongside `--document-private-items`).

GuillaumeGomez commented on src/librustdoc/formats/item_type.rs:
```diff
@@ -58,6 +58,7 @@ pub(crate) enum ItemType {
     TraitAlias = 25,
     // This number is reserved for use in JavaScript
     // Generic = 26,
+    Test = 27,
```
I think it's fine to distinguish functions from test functions, but just in case:

cc @notriddle

GuillaumeGomez commented on src/librustdoc/html/render/mod.rs:
```diff
@@ -419,6 +422,9 @@ impl AllTypes {
         if !self.functions.is_empty() {
             sections.insert(ItemSection::Functions);
         }
+        if !self.tests.is_empty() {
```
I think it'd be better to put the "unit tests" section at the bottom.

GuillaumeGomez commented on src/librustdoc/html/render/mod.rs:
```diff
@@ -2238,6 +2248,7 @@ impl ItemSection {
             Self::Unions => "Unions",
             Self::Enums => "Enums",
             Self::Functions => "Functions",
+            Self::Tests => "Tests",
```
Maybe "Unit tests" instead? Same above.

notriddle replied:
This bikeshed can continue as part of the FCP, I think.

GuillaumeGomez commented on src/librustdoc/html/render/print_item.rs:
```diff
@@ -186,6 +186,7 @@ pub(super) fn print_item(cx: &mut Context<'_>, item: &clean::Item, buf: &mut Buf
             }
         }
         clean::FunctionItem(..) | clean::ForeignFunctionItem(..) => "Function ",
+        clean::TestItem(..) => "Test ",
```
`"Unit test "`

ifxfrancois replied:
We also want to call it on integration tests. If we change this to "Unit Test", it will also appear as Unit test in the integration tests.
With a command like `cargo rustdoc --test lib_test -- -Z unstable-options --document-tests --document-private-items`

GuillaumeGomez commented:
I think it's a good idea. Just one thing missing: an entry in the rustdoc book about the command line flag and that it needs to be used with `--document-private-items` and likely with `--cfg test` as well.

ifxfrancois commented:
> I think it's a good idea. Just one thing missing: an entry in the rustdoc book about the command line flag and that it needs to be used with `--document-private-items` and likely with `--cfg test` as well.

@GuillaumeGomez  I don't know about the correct page where to add it, should I add it to the page "Unstable features" or to the page "Command-line arguments".

GuillaumeGomez commented:
Sorry should have precised. Since the option is unstable, please put it into "Unstable features".

Urgau commented:
> I think it's a good idea. Just one thing missing: an entry in the rustdoc book about the command line flag and that it needs to be used with `--document-private-items` and likely with `--cfg test` as well.

It would be quite easy for rustdoc to automaticly set the `test` cfg when `--document-tests` is passed, in the same vain as done with the `doc` cfg. https://github.com/rust-lang/rust/blob/82d17a4db3a010caa947d49245de63a7ac14accf/src/librustdoc/core.rs#L206-L207

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-17`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/10921553600/job/30313922056) [(plain)](https://github.com/rust-lang/rust/commit/b93d6ba02b06a88a004e2e874593cd7c88db07a0/checks/30313922056/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
------
 > importing cache manifest from ghcr.io/rust-lang/rust-ci-cache:fb97a1d6377f6cf2227825318ca4bbde3889e0c420746f5a03ba46a07e9a862725c26a09b9fc49a0d129ebd75935d3f6cd19acf41cc4267a6846fd4aa574b12c:
------
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-17]
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-17', '--enable-llvm-link-shared', '--set', 'rust.randomize-layout=true', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'change-id=99999999', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-17/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.randomize-layout := True
configure: rust.thin-lto-import-instr-limit := 10
---
  Downloaded boml v0.3.1
   Compiling boml v0.3.1
   Compiling y v0.1.0 (/checkout/compiler/rustc_codegen_gcc/build_system)
    Finished `release` profile [optimized] target(s) in 3.83s
     Running `/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-codegen/x86_64-unknown-linux-gnu/release/y test --use-system-gcc --use-backend gcc --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc --release --mini-tests --std-tests`
Using system GCC
[BUILD] example
[AOT] mini_core_hello_world
/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc/mini_core_hello_world
abc
---
   Compiling rustdoc v0.0.0 (/checkout/src/librustdoc)
error[E0061]: this function takes 3 arguments but 2 arguments were supplied
   --> src/librustdoc/clean/types/tests.rs:74:17
    |
74  |     let cache = Cache::new(false, false);
    |                 ^^^^^^^^^^-------------- argument #3 of type `bool` is missing
note: associated function defined here
   --> src/librustdoc/formats/cache.rs:150:19
    |
    |
150 |     pub(crate) fn new(document_private: bool, document_hidden: bool, document_tests: bool) -> Self {
help: provide the argument
    |
    |
74  |     let cache = Cache::new(false, false, /* bool */);

For more information about this error, try `rustc --explain E0061`.
error: could not compile `rustdoc` (lib test) due to 1 previous error
warning: build failed, waiting for other jobs to finish...
```

</details>

GuillaumeGomez commented:
> > I think it's a good idea. Just one thing missing: an entry in the rustdoc book about the command line flag and that it needs to be used with `--document-private-items` and likely with `--cfg test` as well.
> 
> It would be quite easy for rustdoc to automaticly set the `test` cfg when `--document-tests` is passed, in the same vain as done with the `doc` cfg.
> 
> https://github.com/rust-lang/rust/blob/82d17a4db3a010caa947d49245de63a7ac14accf/src/librustdoc/core.rs#L206-L207

Yes it's easy to do. The question is more: "should we do it?".

ifxfrancois commented on src/librustdoc/core.rs:
```diff
@@ -229,7 +229,8 @@ pub(crate) fn create_config(
         if *document_private { ResolveDocLinks::All } else { ResolveDocLinks::Exported };
-    let test = scrape_examples_options.map(|opts| opts.scrape_tests).unwrap_or(false);
+    let test =
+        scrape_examples_options.map(|opts| opts.scrape_tests).unwrap_or(false) || *document_tests;
```
> > > I think it's a good idea. Just one thing missing: an entry in the rustdoc book about the command line flag and that it needs to be used with `--document-private-items` and likely with `--cfg test` as well.
> > 
> > 
> > It would be quite easy for rustdoc to automaticly set the `test` cfg when `--document-tests` is passed, in the same vain as done with the `doc` cfg.
> > https://github.com/rust-lang/rust/blob/82d17a4db3a010caa947d49245de63a7ac14accf/src/librustdoc/core.rs#L206-L207
> 
> Yes it's easy to do. The question is more: "should we do it?".

In this implementation, ' --cfg test` is already passed automatically here.
It may not be the cleanest way to pass it.

Having it passed automatically is convenient, I think it would be a bit counter-intuitive if the option "document-tests" did not actually document tests except if an additional flag is passed.

GuillaumeGomez replied:
I think it's an important thing to discuss as it might not be obvious for users. I would personally pass `--cfg test` as I wouldn't expect an option to modify the `cfg`s.

jieyouxu replied:
possible relevant MCP https://github.com/rust-lang/compiler-team/issues/785

jieyouxu replied:
FYI, https://github.com/rust-lang/compiler-team/issues/785 has now entered FCP.

notriddle replied:
> Starting with 1.85.0, the test cfg is consider to be a "userspace" config despite being also set by rustc and should be managed by the build system itself.

https://doc.rust-lang.org/nightly/rustc/check-cfg.html

ifxfrancois committed:
Add support to generate documentation for tests

The new option --document-tests is unstable and documented as such.
In order to use it is needed to add `--cfg test` and in case the tests
are not marked public to add `--document-private-items`.
The implementation hide the auto generate main test function and
constants.

bors commented:
:umbrella: The latest upstream changes (presumably #130724) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-18`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/11218121442/job/31181219578) [(plain)](https://github.com/rust-lang/rust/commit/f8de95cf5844b0c3d81573629550d17fc7c431dd/checks/31181219578/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
------
 > importing cache manifest from ghcr.io/rust-lang/rust-ci-cache:20950f2ccee3dff53a038adf5c1cf05231c0b30772617126a5f6478a66316a29cba9aead69f7bb0004886d32c1f8e6287542cf6d25130711c82d16a66201d4fe:
------
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-18]
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-18', '--enable-llvm-link-shared', '--set', 'rust.randomize-layout=true', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'change-id=99999999', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-18/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.randomize-layout := True
configure: rust.thin-lto-import-instr-limit := 10
---
  Downloaded boml v0.3.1
   Compiling boml v0.3.1
   Compiling y v0.1.0 (/checkout/compiler/rustc_codegen_gcc/build_system)
    Finished `release` profile [optimized] target(s) in 3.72s
     Running `/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-codegen/x86_64-unknown-linux-gnu/release/y test --use-system-gcc --use-backend gcc --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc --release --mini-tests --std-tests`
Using system GCC
[BUILD] example
[AOT] mini_core_hello_world
/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc/mini_core_hello_world
abc
---
   Compiling rustdoc v0.0.0 (/checkout/src/librustdoc)
error[E0061]: this function takes 3 arguments but 2 arguments were supplied
   --> src/librustdoc/clean/types/tests.rs:74:17
    |
74  |     let cache = Cache::new(false, false);
    |                 ^^^^^^^^^^-------------- argument #3 of type `bool` is missing
note: associated function defined here
   --> src/librustdoc/formats/cache.rs:150:19
    |
    |
150 |     pub(crate) fn new(document_private: bool, document_hidden: bool, document_tests: bool) -> Self {
help: provide the argument
    |
    |
74  |     let cache = Cache::new(false, false, /* bool */);

For more information about this error, try `rustc --explain E0061`.
error: could not compile `rustdoc` (lib test) due to 1 previous error
warning: build failed, waiting for other jobs to finish...
```

</details>

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-18`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/11239996575/job/31248341150) [(plain)](https://github.com/rust-lang/rust/commit/fa3e84ff00bad541d9bb1261674716248f2db7b3/checks/31248341150/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
------
 > importing cache manifest from ghcr.io/rust-lang/rust-ci-cache:20950f2ccee3dff53a038adf5c1cf05231c0b30772617126a5f6478a66316a29cba9aead69f7bb0004886d32c1f8e6287542cf6d25130711c82d16a66201d4fe:
------
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-18]
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-18', '--enable-llvm-link-shared', '--set', 'rust.randomize-layout=true', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'change-id=99999999', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-18/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.randomize-layout := True
configure: rust.thin-lto-import-instr-limit := 10
---
  Downloaded boml v0.3.1
   Compiling boml v0.3.1
   Compiling y v0.1.0 (/checkout/compiler/rustc_codegen_gcc/build_system)
    Finished `release` profile [optimized] target(s) in 3.89s
     Running `/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-codegen/x86_64-unknown-linux-gnu/release/y test --use-system-gcc --use-backend gcc --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc --release --mini-tests --std-tests`
Using system GCC
[BUILD] example
[AOT] mini_core_hello_world
/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc/mini_core_hello_world
abc
---
---- [run-make] tests/run-make/rustdoc-default-output stdout ----

error: rmake recipe failed to complete
status: exit status: 101
command: cd "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/rustdoc-default-output/rmake_out" && env -u RUSTFLAGS AR="ar" CARGO="/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools-bin/cargo" CC="cc" CC_DEFAULT_FLAGS="-ffunction-sections -fdata-sections -fPIC -m64" CXX="c++" CXX_DEFAULT_FLAGS="-ffunction-sections -fdata-sections -fPIC -m64" HOST_RPATH_DIR="/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib" LD_LIBRARY_PATH="/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-bootstrap-tools/x86_64-unknown-linux-gnu/release/deps:/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/lib:/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools-bin:/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" LD_LIB_PATH_ENVVAR="LD_LIBRARY_PATH" LLVM_BIN_DIR="/usr/lib/llvm-18/bin" LLVM_COMPONENTS="aarch64 aarch64asmparser aarch64codegen aarch64desc aarch64disassembler aarch64info aarch64utils aggressiveinstcombine all all-targets amdgpu amdgpuasmparser amdgpucodegen amdgpudesc amdgpudisassembler amdgpuinfo amdgputargetmca amdgpuutils analysis arm armasmparser armcodegen armdesc armdisassembler arminfo armutils asmparser asmprinter avr avrasmparser avrcodegen avrdesc avrdisassembler avrinfo binaryformat bitreader bitstreamreader bitwriter bpf bpfasmparser bpfcodegen bpfdesc bpfdisassembler bpfinfo cfguard codegen codegentypes core coroutines coverage debuginfobtf debuginfocodeview debuginfodwarf debuginfogsym debuginfologicalview debuginfomsf debuginfopdb demangle dlltooldriver dwarflinker dwarflinkerclassic dwarflinkerparallel dwp engine executionengine extensions filecheck frontenddriver frontendhlsl frontendoffloading frontendopenacc frontendopenmp fuzzercli fuzzmutate globalisel hexagon hexagonasmparser hexagoncodegen hexagondesc hexagondisassembler hexagoninfo hipstdpar instcombine instrumentation interfacestub interpreter ipo irprinter irreader jitlink lanai lanaiasmparser lanaicodegen lanaidesc lanaidisassembler lanaiinfo libdriver lineeditor linker loongarch loongarchasmparser loongarchcodegen loongarchdesc loongarchdisassembler loongarchinfo lto m68k m68kasmparser m68kcodegen m68kdesc m68kdisassembler m68kinfo mc mca mcdisassembler mcjit mcparser mips mipsasmparser mipscodegen mipsdesc mipsdisassembler mipsinfo mirparser msp430 msp430asmparser msp430codegen msp430desc msp430disassembler msp430info native nativecodegen nvptx nvptxcodegen nvptxdesc nvptxinfo objcarcopts objcopy object objectyaml option orcdebugging orcjit orcshared orctargetprocess passes perfjitevents powerpc powerpcasmparser powerpccodegen powerpcdesc powerpcdisassembler powerpcinfo profiledata remarks riscv riscvasmparser riscvcodegen riscvdesc riscvdisassembler riscvinfo riscvtargetmca runtimedyld scalaropts selectiondag sparc sparcasmparser sparccodegen sparcdesc sparcdisassembler sparcinfo support symbolize systemz systemzasmparser systemzcodegen systemzdesc systemzdisassembler systemzinfo tablegen target targetparser textapi textapibinaryreader transformutils ve veasmparser vecodegen vectorize vedesc vedisassembler veinfo webassembly webassemblyasmparser webassemblycodegen webassemblydesc webassemblydisassembler webassemblyinfo webassemblyutils windowsdriver windowsmanifest x86 x86asmparser x86codegen x86desc x86disassembler x86info x86targetmca xcore xcorecodegen xcoredesc xcoredisassembler xcoreinfo xray xtensa xtensaasmparser xtensacodegen xtensadesc xtensadisassembler xtensainfo" LLVM_FILECHECK="/usr/lib/llvm-18/bin/FileCheck" NODE="/usr/bin/node" PYTHON="/usr/bin/python3" RUSTC="/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" RUSTDOC="/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustdoc" SOURCE_ROOT="/checkout" TARGET="x86_64-unknown-linux-gnu" TARGET_RPATH_DIR="/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" TARGET_RPATH_ENV="/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/rustdoc-default-output/rmake_out:/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-bootstrap-tools/x86_64-unknown-linux-gnu/release/deps:/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/lib" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/rustdoc-default-output/rmake"
--- stderr -------------------------------
thread 'main' panicked at /checkout/tests/run-make/rustdoc-default-output/rmake.rs:15:10:
test failed: `output-default.stdout` is different from `actual`

```

</details>

bors commented:
:umbrella: The latest upstream changes (presumably #131573) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/11323129220/job/31485172355) [(plain)](https://github.com/rust-lang/rust/commit/8e4c7a9d3bacf411e072f0de9d73e9d8380393db/checks/31485172355/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---
#13 2.863 Building wheels for collected packages: reuse
#13 2.864   Building wheel for reuse (pyproject.toml): started
#13 3.111   Building wheel for reuse (pyproject.toml): finished with status 'done'
#13 3.112   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132720 sha256=026f3bb0f1aa8090b861fd0a0939cb1a782396d84c8aab7875096557d637a0f6
#13 3.112   Stored in directory: /tmp/pip-ephem-wheel-cache-778c8ciq/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#13 3.115 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#13 3.506 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#13 3.506 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#13 4.034 Collecting virtualenv
#13 4.034 Collecting virtualenv
#13 4.089   Downloading virtualenv-20.26.6-py3-none-any.whl (6.0 MB)
#13 4.304      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 6.0/6.0 MB 28.1 MB/s eta 0:00:00
#13 4.365 Collecting platformdirs<5,>=3.9.1
#13 4.372   Downloading platformdirs-4.3.6-py3-none-any.whl (18 kB)
#13 4.392 Collecting distlib<1,>=0.3.7
#13 4.409      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 469.0/469.0 KB 58.9 MB/s eta 0:00:00
#13 4.446 Collecting filelock<4,>=3.12.2
#13 4.453   Downloading filelock-3.16.1-py3-none-any.whl (16 kB)
#13 4.453   Downloading filelock-3.16.1-py3-none-any.whl (16 kB)
#13 4.532 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#13 4.720 Successfully installed distlib-0.3.9 filelock-3.16.1 platformdirs-4.3.6 virtualenv-20.26.6
#13 DONE 4.8s

#14 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#14 DONE 0.0s
---
DirectMap4k:      225216 kB
DirectMap2M:     9211904 kB
DirectMap1G:     9437184 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
    Finished `dev` profile [unoptimized] target(s) in 0.04s
##[endgroup]
WARNING: `rust.download-rustc` is enabled. The `rust.channel` option will be overridden by the CI rustc's channel.
downloading https://static.rust-lang.org/dist/2024-09-22/rustfmt-nightly-x86_64-unknown-linux-gnu.tar.xz
---
fmt check
Diff in /checkout/src/librustdoc/json/conversions.rs:10:
 use rustc_hir::def_id::DefId;
 use rustc_metadata::rendered_const;
 use rustc_middle::{bug, ty};
-use rustc_span::{sym, Pos, Symbol};
+use rustc_span::{Pos, Symbol, sym};
 use rustc_target::spec::abi::Abi as RustcAbi;
 
Diff in /checkout/src/librustdoc/json/conversions.rs:17:
 use super::FullItemId;
 use crate::clean::{self, ItemId};
 use crate::clean::{self, ItemId};
-use crate::formats::item_type::ItemType;
 use crate::formats::FormatRenderer;
+use crate::formats::item_type::ItemType;
 use crate::json::JsonRenderer;
 use crate::passes::collect_intra_doc_links::UrlFragment;
 
fmt error: Running `"/checkout/obj/build/x86_64-unknown-linux-gnu/rustfmt/bin/rustfmt" "--config-path" "/checkout" "--edition" "2021" "--unstable-features" "--skip-children" "--check" "/checkout/tests/assembly/nvptx-arch-emit-asm.rs" "/checkout/tests/assembly/sparc-struct-abi.rs" "/checkout/tests/assembly/issue-83585-small-pod-struct-equality.rs" "/checkout/tests/assembly/x86-return-float.rs" "/checkout/tests/assembly/align_offset.rs" "/checkout/tests/assembly/cmse.rs" "/checkout/tests/assembly/nvptx-linking-binary.rs" "/checkout/tests/assembly/pie-relocation-model.rs" "/checkout/tests/assembly/nvptx-c-abi-arg-v7.rs" "/checkout/tests/assembly/nvptx-arch-default.rs" "/checkout/tests/assembly/stack-protector/stack-protector-heuristics-effect-windows-32bit.rs" "/checkout/tests/assembly/stack-protector/stack-protector-target-support.rs" "/checkout/tests/assembly/stack-protector/stack-protector-heuristics-effect.rs" "/checkout/tests/assembly/stack-protector/stack-protector-heuristics-effect-windows-64bit.rs" "/checkout/tests/assembly/nvptx-kernel-abi/nvptx-kernel-args-abi-v7.rs" "/checkout/tests/assembly/libs/issue-115339-zip-arrays.rs" "/checkout/tests/assembly/aarch64-naked-fn-no-bti-prolog.rs" "/checkout/tests/assembly/strict_provenance.rs" "/checkout/tests/assembly/dwarf4.rs" "/checkout/tests/assembly/x86_64-function-return.rs" "/checkout/tests/assembly/x86_64-fortanix-unknown-sgx-lvi-generic-ret.rs" "/checkout/tests/assembly/slice-is_ascii.rs" "/checkout/tests/assembly/nvptx-internalizing.rs" "/checkout/tests/assembly/aarch64-pointer-auth.rs" "/checkout/tests/assembly/nvptx-c-abi-ret-v7.rs" "/checkout/tests/assembly/nvptx-linking-cdylib.rs" "/checkout/tests/assembly/closure-inherit-target-feature.rs" "/checkout/tests/assembly/auxiliary/non-inline-dependency.rs" "/checkout/tests/assembly/auxiliary/breakpoint-panic-handler.rs" "/checkout/tests/assembly/pic-relocation-model.rs" "/checkout/tests/assembly/target-feature-multiple.rs" "/checkout/tests/assembly/is_aligned.rs" "/checkout/tests/assembly/x86_64-fortanix-unknown-sgx-lvi-generic-load.rs" "/checkout/tests/assembly/x86_64-sse_crc.rs" "/checkout/tests/assembly/simd-intrinsic-mask-store.rs" "/checkout/tests/assembly/static-relocation-model.rs" "/checkout/tests/assembly/simd-intrinsic-mask-reduce.rs" "/checkout/tests/assembly/dwarf5.rs" "/checkout/tests/assembly/x86_64-naked-fn-no-cet-prolog.rs" "/checkout/tests/assembly/targets/targets-elf.rs" "/checkout/tests/assembly/targets/targets-nvptx.rs" "/checkout/tests/assembly/targets/targets-pe.rs" "/checkout/tests/assembly/targets/targets-macho.rs" "/checkout/tests/assembly/simd-bitmask.rs" "/checkout/tests/assembly/stack-probes.rs" "/checkout/tests/assembly/niche-prefer-zero.rs" "/checkout/tests/assembly/x86_64-array-pair-load-store-merge.rs" "/checkout/src/librustdoc/clean/render_macro_matchers.rs" "/checkout/src/librustdoc/clean/auto_trait.rs" "/checkout/src/librustdoc/clean/utils.rs" "/checkout/src/librustdoc/clean/cfg/tests.rs" "/checkout/src/librustdoc/clean/mod.rs" "/checkout/src/librustdoc/markdown.rs" "/checkout/src/librustdoc/json/import_finder.rs" "/checkout/src/librustdoc/json/conversions.rs" "/checkout/src/librustdoc/json/mod.rs" "/checkout/src/librustdoc/doctest.rs" "/checkout/src/librustdoc/scrape_examples.rs" "/checkout/src/librustdoc/docfs.rs" "/checkout/src/librustdoc/fold.rs" "/checkout/src/librustdoc/config.rs" "/checkout/src/librustdoc/core.rs" "/checkout/src/librustdoc/externalfiles.rs" "/checkout/tests/assembly/x86_64-floating-point-clamp.rs"` failed.
If you're running `tidy`, try again with `--bless`. Or, if you just want to format code, run `./x.py fmt` instead.
  local time: Mon Oct 14 07:53:11 UTC 2024
  network time: Mon, 14 Oct 2024 07:53:11 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

bors commented:
:umbrella: The latest upstream changes (presumably #132815) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-18`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/11775843137/job/32797073589) [(plain)](https://github.com/rust-lang/rust/commit/a9eec750e8e598e539b37a371834531a6377dae0/checks/32797073589/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
------
 > importing cache manifest from ghcr.io/rust-lang/rust-ci-cache:c32c805632780b5c1de330e3f44561b336c2efe163bc0990acb392390157a8e1d9f855d75914a239aa40c49d77f4a837247d05d2f8d46f554b98e1f46712a3e3:
------
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-18]
debug: `DISABLE_CI_RUSTC_IF_INCOMPATIBLE` configured.
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-18', '--enable-llvm-link-shared', '--set', 'rust.randomize-layout=true', '--set', 'rust.thin-lto-import-instr-limit=10', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-18/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.randomize-layout := True
configure: rust.thin-lto-import-instr-limit := 10
---
  Downloaded boml v0.3.1
   Compiling boml v0.3.1
   Compiling y v0.1.0 (/checkout/compiler/rustc_codegen_gcc/build_system)
    Finished `release` profile [optimized] target(s) in 3.87s
     Running `/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-codegen/x86_64-unknown-linux-gnu/release/y test --use-system-gcc --use-backend gcc --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc --release --mini-tests --std-tests`
Using system GCC
[BUILD] example
[AOT] mini_core_hello_world
/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc/mini_core_hello_world
abc
---
---- [run-make] tests/run-make/rustdoc-default-output stdout ----

error: rmake recipe failed to complete
status: exit status: 101
command: cd "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/rustdoc-default-output/rmake_out" && env -u RUSTFLAGS AR="ar" CARGO="/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools-bin/cargo" CC="cc" CC_DEFAULT_FLAGS="-ffunction-sections -fdata-sections -fPIC -m64" CXX="c++" CXX_DEFAULT_FLAGS="-ffunction-sections -fdata-sections -fPIC -m64" HOST_RPATH_DIR="/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib" LD_LIBRARY_PATH="/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-bootstrap-tools/x86_64-unknown-linux-gnu/release/deps:/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/lib:/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools-bin:/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" LD_LIB_PATH_ENVVAR="LD_LIBRARY_PATH" LLVM_BIN_DIR="/usr/lib/llvm-18/bin" LLVM_COMPONENTS="aarch64 aarch64asmparser aarch64codegen aarch64desc aarch64disassembler aarch64info aarch64utils aggressiveinstcombine all all-targets amdgpu amdgpuasmparser amdgpucodegen amdgpudesc amdgpudisassembler amdgpuinfo amdgputargetmca amdgpuutils analysis arm armasmparser armcodegen armdesc armdisassembler arminfo armutils asmparser asmprinter avr avrasmparser avrcodegen avrdesc avrdisassembler avrinfo binaryformat bitreader bitstreamreader bitwriter bpf bpfasmparser bpfcodegen bpfdesc bpfdisassembler bpfinfo cfguard codegen codegentypes core coroutines coverage debuginfobtf debuginfocodeview debuginfodwarf debuginfogsym debuginfologicalview debuginfomsf debuginfopdb demangle dlltooldriver dwarflinker dwarflinkerclassic dwarflinkerparallel dwp engine executionengine extensions filecheck frontenddriver frontendhlsl frontendoffloading frontendopenacc frontendopenmp fuzzercli fuzzmutate globalisel hexagon hexagonasmparser hexagoncodegen hexagondesc hexagondisassembler hexagoninfo hipstdpar instcombine instrumentation interfacestub interpreter ipo irprinter irreader jitlink lanai lanaiasmparser lanaicodegen lanaidesc lanaidisassembler lanaiinfo libdriver lineeditor linker loongarch loongarchasmparser loongarchcodegen loongarchdesc loongarchdisassembler loongarchinfo lto m68k m68kasmparser m68kcodegen m68kdesc m68kdisassembler m68kinfo mc mca mcdisassembler mcjit mcparser mips mipsasmparser mipscodegen mipsdesc mipsdisassembler mipsinfo mirparser msp430 msp430asmparser msp430codegen msp430desc msp430disassembler msp430info native nativecodegen nvptx nvptxcodegen nvptxdesc nvptxinfo objcarcopts objcopy object objectyaml option orcdebugging orcjit orcshared orctargetprocess passes perfjitevents powerpc powerpcasmparser powerpccodegen powerpcdesc powerpcdisassembler powerpcinfo profiledata remarks riscv riscvasmparser riscvcodegen riscvdesc riscvdisassembler riscvinfo riscvtargetmca runtimedyld scalaropts selectiondag sparc sparcasmparser sparccodegen sparcdesc sparcdisassembler sparcinfo support symbolize systemz systemzasmparser systemzcodegen systemzdesc systemzdisassembler systemzinfo tablegen target targetparser textapi textapibinaryreader transformutils ve veasmparser vecodegen vectorize vedesc vedisassembler veinfo webassembly webassemblyasmparser webassemblycodegen webassemblydesc webassemblydisassembler webassemblyinfo webassemblyutils windowsdriver windowsmanifest x86 x86asmparser x86codegen x86desc x86disassembler x86info x86targetmca xcore xcorecodegen xcoredesc xcoredisassembler xcoreinfo xray xtensa xtensaasmparser xtensacodegen xtensadesc xtensadisassembler xtensainfo" LLVM_FILECHECK="/usr/lib/llvm-18/bin/FileCheck" NODE="/usr/bin/node" PYTHON="/usr/bin/python3" RUSTC="/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" RUSTDOC="/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustdoc" SOURCE_ROOT="/checkout" TARGET="x86_64-unknown-linux-gnu" TARGET_RPATH_DIR="/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" TARGET_RPATH_ENV="/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/rustdoc-default-output/rmake_out:/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-bootstrap-tools/x86_64-unknown-linux-gnu/release/deps:/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/lib" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/rustdoc-default-output/rmake"
--- stderr -------------------------------
thread 'main' panicked at /checkout/tests/run-make/rustdoc-default-output/rmake.rs:15:10:
test failed: `output-default.stdout` is different from `actual`

---
-        --document-tests 
-                        Generate documentation for tests
         --with-examples path to function call information (for displaying examples in the documentation)
                         
         --merge none|shared|finalize
                         more information
         --html-no-source 
                         Disable HTML source code pages generation
+        --document-tests 
+        --document-tests 
+                        Generate documentation for tests
 
     @path               Read newline separated options from `path`

note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
------------------------------------------

```

</details>

bors commented:
:umbrella: The latest upstream changes (presumably #133193) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

lucagladiator commented:
I just fixed the merge conflicts and the checks are now green again.
I tried to address the comments that have been brought up so far and there have been no comments lately.
Can we move on and merge this? Or should we wait for more comments?

Dylan-DPC commented:
> Can we move on and merge this? Or should we wait for more comments?
This will need another review by the reviewer.

ifxfrancois commented:
> > Can we move on and merge this? Or should we wait for more comments?
> 
> This will need another review by the reviewer.

I think that if the reviewers agree, it could be merged as is.

bors commented:
:umbrella: The latest upstream changes (presumably #134550) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

bors commented:
:umbrella: The latest upstream changes (presumably #134590) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

bors commented:
:umbrella: The latest upstream changes (presumably #135031) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

bors commented:
:umbrella: The latest upstream changes (presumably #135947) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

bors commented:
:umbrella: The latest upstream changes (presumably #137425) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

bors commented:
:umbrella: The latest upstream changes (presumably #137914) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

bors commented:
:umbrella: The latest upstream changes (presumably #138021) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

bors commented:
:umbrella: The latest upstream changes (presumably #138630) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

notriddle commented on src/librustdoc/visit_ast.rs:
```diff
@@ -506,11 +523,35 @@ impl<'a, 'tcx> RustdocVisitor<'a, 'tcx> {
+                    if self.cx.cache.document_tests
+                        && let TyKind::Path(rustc_hir::QPath::Resolved(_, path)) = ty.kind
+                    {
+                        let path: String = path
+                            .segments
+                            .iter()
+                            .map(|s| {
+                                if s.ident.name == kw::PathRoot {
+                                    ""
+                                } else {
+                                    s.ident.name.as_str()
+                                }
+                            })
+                            .intersperse("::")
+                            .collect();
+                        if path == "test::TestDescAndFn" {
```
```suggestion
                        let path_names = path
                            .segments
                            .iter()
                            .map(|s| {
                                s.ident.name
                            });
                        if path_names.eq([sym::test, sym::TestDescAndFn]) {
```

I think, to follow this suggestion, you'd need to add TestDescAndFn to <https://doc.rust-lang.org/nightly/nightly-rustc/src/rustc_span/symbol.rs.html#169>

notriddle commented:
@rfcbot fcp merge

Manishearth commented:
@rfcbot concern motivation

> During the development of embedded software, Quality or certifications teams need a quick overview of the tests of a project. 

I'm not really convinced of this motivation. This seems perfect for a domain specific tool, perhaps one consuming rustdoc JSON output.


If we are going down this path I would expect a lot more work on what test documentation might look like, how it should be presented, etc, and probably an RFC.

My subsequent two concerns somewhat reflect my issue with this motivation; as something written with a rather domain specific motivation, it overindexes on a particular way of testing Rust code.

@rfcbot concern doctests

What about doctests? Different projects have different conventions here: some projects have doctests as their primary method of testing, some have unit/integration tests that cover everything important and just run doctests as a way of keeping the examples working.


@rfcbot concern integration-tests

What about integration tests? These are not handled here at all.

bors commented:
:umbrella: The latest upstream changes (presumably #139881) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

notriddle commented:
Why would doctests be a problem? They’re always included in the documentation.

Manishearth commented:
> Why would doctests be a problem? They’re always included in the documentation.

Because the stated goal is about having a "quick overview" of the tests here. Doctests don't show up that way.

Plus, with such a motivation, you probably want to document your tests, and doctests cannot be documented. This is a situation of tension between the dual purposes of doctests: as tests, and as examples.

lucagladiator commented:
Thank you very much for the feedback!

I agree, that the motivation is quite narrow at the moment.
There might be more use cases, but we see it as immediately useful to industries seeking certification of their software (e.g. automotive).

For us JSON output that we can process with other tools would be fine, but that would still need a change in rustdoc to generate this JSON output.

Regarding your other concerns:

- integration tests work in theory, but they are compiled in separate compilation units. Because of this, our current implementation can only generate documentation for a single test in isolation. This is not really useful and would require some other solution or tooling to combine the documentation. Maybe this could be achieved with cargo, but we did not look into that (yet).
- doctests are more complicated. As you already said, there is currently no way for them to be documented. The only thing that comes to my mind at the moment would be to use `#![doc = "some documentation"]` inside the doctest, but I am not very convinced by my own idea. If we go for a more general solution we should definitively check for better ways to handle doctests.

How would you suggest we move forward with this topic? Should we adapt this PR so it just generates the JSON output but no HTML? Or would you say we should create an RFC first either way?

Manishearth commented:
Rustdoc already produces json output (`--output-format json`). You should be able to get the output you need by pairing that with `--document-private-items`.

I'm noticing two issues:

 - It doesn't seem to include attributes on functions (this should be fixed; rustdoc-HTML does include e.g. the `no_mangle` attribute in its displayed output)
 - Even if it did, it would filter out `#[test]`, due to [this code](https://github.com/rust-lang/rust/blob/883f9f72e87ccb6838d528d8158ea6323baacc65/src/librustdoc/clean/types.rs#L763). Given that that attribute is mostly only on private items I think it's fine to have that be one of the displayed doc attributes.

So for next steps I would recommend experimenting with the existing tooling, and trying to patch up gaps in it.

lucagladiator commented:
I just tried to generate JSON output for with tests but I could not get it to work. I also tried using rustdoc with `--cfg test` which included test modules, but not the tests themselves.

I believe the problem is that the HIR only contains tests if `--test` is passed to rustc. As far as I can see, rustdoc only passes `--test` to rustc [when scraping examples from tests](https://github.com/rust-lang/rust/blob/master/src/librustdoc/core.rs#L253). In your opinion, what would be the best way to allow passing `--test` to rustc through rustdoc? A new CLI flag or some other approach?

> * It doesn't seem to include attributes on functions (this should be fixed; rustdoc-HTML does include e.g. the `no_mangle` attribute in its displayed output)
> * Even if it did, it would filter out `#[test]`, due to [this code](https://github.com/rust-lang/rust/blob/883f9f72e87ccb6838d528d8158ea6323baacc65/src/librustdoc/clean/types.rs#L763). Given that that attribute is mostly only on private items I think it's fine to have that be one of the displayed doc attributes.

Looking at the HIR generated by rustc (`cargo rustc -- -Z unpretty=hir --test`) it looks like rustc is doing quite a lot of transformations on the tests. In particular, the `#[test]` does not appear in the HIR. This would mean the tooling working with the JSON output would have to do something similar to what this PR is doing with regards to recovering the tests from `TestDescAndFn` constants.

GuillaumeGomez commented:
No, I think your problem is that private items are stripped by default. Try with `--document-private-items` too.

lucagladiator commented:
> No, I think your problem is that private items are stripped by default. Try with `--document-private-items` too.

I did try with `--document-private-items`, `--cfg test` and `--test`. None of those produces a JSON file which contains any mention of my example test (When passing `--test` no JSON file is generated at all).

My current test setup is a crate with a lib.rs which just contains the following test:
```rust
/// A test with documentation.
#[test]
fn some_documented_test() {
    assert_eq!(2 + 2, 4);
}
```
I am running `rustdoc +nightly --document-private-items -Z unstable-options --output-format json src/lib.rs`
The resulting JSON does not contain `some_documented_test` at all.

GuillaumeGomez commented:
I see, it seems like a new option is needed to make it work then.

Manishearth commented:
No, I don't see why a new option is needed, there is no reason that `document-private-items` cannot have this behavior already.

Nothing here needs new functionality, it just needs tweaks to existing functionality.

GuillaumeGomez commented:
We mean that if there is a visible `#[test]` function, it should be listed.

GuillaumeGomez commented:
I don't see why? Generally they're put behind `cfg(test)`. If not, then I suppose people will discover that when documenting wth `--document-private-items`.
